service/src/main/java/org/jboss/aerogear/unifiedpush/service/impl/ClientInstallationServiceImpl.java

```diff
@@ -83,10 +83,17 @@ public void addInstallation(Variant variant, Installation entity) {
 
     private void mergeCategories(Installation entity) {
         if (entity.getCategories() != null) {
-            List<String> categoryNames = convertToNames(entity.getCategories());
-            final List<Category> categories = categoryDao.findByNames(categoryNames);
-            entity.getCategories().removeAll(categories);
-            entity.getCategories().addAll(categories);
+
+            if(!entity.getCategories().isEmpty()) {
+                final List<String> categoryNames = convertToNames(entity.getCategories());
+                final List<Category> categories = categoryDao.findByNames(categoryNames);
+                entity.getCategories().removeAll(categories);
+                entity.getCategories().addAll(categories);
```

Comment:
a list doesn't have a straight forward replace method adding a comment makes the most sense I guess something like
```java
//replace json deserialised categories with their persistent counter parts see Category.equals
```